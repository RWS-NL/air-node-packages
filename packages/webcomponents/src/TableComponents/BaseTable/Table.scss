@import '../../styles/vars';

.table {
  margin: 0;
  max-width: 100vw;

  &TopPagination {
    margin-bottom: 0.5vw;
  }
}
