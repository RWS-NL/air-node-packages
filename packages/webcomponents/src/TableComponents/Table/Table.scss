@import "../../styles/vars";

.table {
  margin: 0 5vw 1vw;
  max-width: 90vw;

  &TopPagination {
    margin-bottom: 0.5vw;
  }
}
