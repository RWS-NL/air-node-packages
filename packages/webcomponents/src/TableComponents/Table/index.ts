export { default as Table, TableQAs, TableProps } from './Table';