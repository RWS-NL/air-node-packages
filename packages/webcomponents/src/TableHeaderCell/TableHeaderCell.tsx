import MUITableCell, { TableCellProps as MUITableCellProps } from '@material-ui/core/TableCell';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import classnames from 'classnames';
import React, { FC, memo } from 'react';
import { customCss, dataQa, label } from '../constants';
import Tooltip, { TooltipProps } from '../Tooltip/Tooltip';
import css from './TableHeaderCell.scss';

export interface TableHeaderProps {
  /** Label for this table header */
  label: string;
  /** Whether this table header represents a column of numeric values */
  numeric?: boolean;
  /** Whether the additional padding for this table header should be disabled */
  disablePadding?: boolean;
  /** Whether this header is for a cell that contains an action button */
  isActionButtonCell?: boolean;
}

export interface TableHeaderCellProps extends MUITableCellProps {
  /** The header data */
  header: TableHeaderProps;
  /** The property to order by */
  orderby: string;
  /** The order direction for the header */
  order: 'desc' | 'asc';
  /** The label for the tooltip */
  tooltiplabel: label;
  /** The placement of the tooltip */
  tooltipplacement: TooltipProps['placement'];
  /** The function triggered when a sort request is made */
  onRequestSort: (property: string) => void;
  /** Data-qa tag to apply to the search bar and input element */
  'data-qa'?: dataQa;
  /** Custom CSS classes to pass to the button */
  customclasses?: customCss;
  /** Whether this header is for a cell that contains an action button */
  isActionButtonCell?: boolean;
}

/** Creates a table header cell using pre-defined Rijkswaterstaat styling */
export const TableHeaderCell: FC<TableHeaderCellProps> = props => {
  const renderTableHeaderCell = () => {
    if (props.isActionButtonCell) {
      return (
        <TableSortLabel
          data-qa={`tableSortLabel_${props.header.label}`}
          active={false}
          hideSortIcon
          onClick={() => undefined}
        >
          {props.header.label}
        </TableSortLabel>
      );
    }

    return (
      <Tooltip title={props.tooltiplabel} placement={props.tooltipplacement || 'top'}
        data-qa={`table-header-${props['data-qa']}`}>
        <TableSortLabel
          data-qa={`tableSortLabel_${props.header.label}`}
          classes={{icon: css.sortIcon}}
          active={props.orderby === props.header.label}
          direction={props.order}
          onClick={() => props.onRequestSort(props.header.label)}
        >
          {props.header.label}
        </TableSortLabel>
      </Tooltip>
    );
  };

  return (
    <MUITableCell
      data-qa={props['data-qa']}
      align={props.header.numeric ? 'right' : 'left'}
      padding={props.header.disablePadding ? 'none' : 'default'}
      sortDirection={props.orderby === props.header.label ? props.order : false}
      className={classnames(css.tableCells, props.customclasses)}>
      {renderTableHeaderCell()}
    </MUITableCell>
  );
};

export default memo(TableHeaderCell);