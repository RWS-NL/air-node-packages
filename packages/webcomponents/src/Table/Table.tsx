import MUITable, { TableProps as MUITableProps } from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableHead from '@material-ui/core/TableHead';
import { TablePaginationProps as MUITablePaginationProps } from '@material-ui/core/TablePagination';
import TableRow from '@material-ui/core/TableRow';
import classnames from 'classnames';
import React, { FC, Fragment, ReactNode, memo } from 'react';
import css from './Table.scss';
import TableHeaderCell, { TableHeaderCellProps, TableHeaderProps } from '../TableHeaderCell/TableHeaderCell';
import TablePagination, { TablePaginationProps } from '../TablePagination/TablePagination';
import TableToolbar, { TableToolbarProps } from '../TableToolbar/TableToolbar';
import { dataQa } from '../constants';

export type TableLabels =
  & Pick<TableToolbarProps, 'searchplaceholderlabel'>
  & Pick<TablePaginationProps, 'labelRowsPerPage' | 'labelPaginationOf'>
  & Pick<TableHeaderCellProps, 'tooltiplabel'>;

export interface TableQAs {
  /** Data-qa applied to the root table */
  table: dataQa;
  /** Data-qa applied to the pagination components */
  pagination: dataQa;
  /** Data-qa applied to the toolbar */
  toolbar: dataQa;
  /** Data-qa applied to the header wrapper */
  header: dataQa;
  /** Data-qa applied to the header row */
  headerRow: dataQa;
  /** Data-qa applied to each header cell */
  headerCell: dataQa;
  /** Data-qa applied to the table body wrapper */
  tableBody: dataQa;
}

export interface TableCustomClasses {
  /** Custom class(es) for table */
  table?: string[];
  /** Custom class(es) for table toolbar */
  tableToolbar?: string[];
  /** Custom class(es) for table paginations */
  tablePagination?: string[];
  /** Custom class(es) for table pagination actions */
  tablePaginationActions?: string[];
  /** Custom class(es) for table header */
  tableHeader?: string[];
  /** Custom class(es) for table header row */
  tableHeaderRow?: string[];
  /** Custom class(es) for table header cell */
  tableHeaderCell?: string[];
  /** Custom class(es) for table body */
  tableBody?: string[];
  /** Custom class(es) for table toolbar searchbar */
  tableSearchbar?: string[];
}

export interface TableProps extends
  Pick<TableToolbarProps, 'onsearchclear' | 'onsearchinput' | 'paperElevation' | 'extraIcons'>,
  Pick<TableHeaderCellProps, 'onRequestSort' | 'orderby' | 'order' | 'tooltipplacement'>,
  Pick<MUITablePaginationProps, 'rowsPerPage' | 'rowsPerPageOptions' | 'page' | 'onChangePage' | 'onChangeRowsPerPage'>,
  MUITableProps {
  /** Headers for the table */
  headers: TableHeaderProps[];
  /** Map that defines how the headers are sorted, either by key or function when targeting nested properties */
  headermapping: Map<string, string>;
  /** The amount of rows for the table (generally the length of the data) */
  rowcount: number;
  /** Labels used in the table */
  labels: TableLabels;
  /** Data-qa's used in the table */
  tableqas: TableQAs;
  /** Content for the table */
  tablebodycontent: ReactNode;
  /** Customization CSS classes for table components */
  tablecss?: TableCustomClasses;
  /** Control whether the top bar pagination should be shown */
  showTopPagination?: boolean;
  /** Control whether the bottom bar pagination should be shown */
  showBottomPagination?: boolean;
}

/** Creates a table using pre-defined Rijkswaterstaat styling */
export const Table: FC<TableProps> = props => {
  const addCustomClasses = (component: keyof TableCustomClasses, baseClass?: string): string[] => {
    const classes: string[] = [];
    if (baseClass) classes.push(baseClass);
    if (props.tablecss && props.tablecss[component]) classes.push(...props.tablecss[component]);

    return classes;
  };

  const renderTablePagination = (customClasses: string) => {
    return (
      <TablePagination
        labelRowsPerPage={props.labels.labelRowsPerPage}
        labelPaginationOf={props.labels.labelPaginationOf}
        rowsPerPageOptions={props.rowsPerPageOptions}
        rowsPerPage={props.rowsPerPage}
        page={props.page}
        count={props.rowcount}
        onChangePage={props.onChangePage}
        onChangeRowsPerPage={props.onChangeRowsPerPage}
        customClasses={customClasses}
        data-qa={props.tableqas.pagination}
      />
    );
  };

  return (
    <Fragment>
      <TableToolbar
        searchplaceholderlabel={props.labels.searchplaceholderlabel}
        onsearchinput={props.onsearchinput}
        onsearchclear={props.onsearchclear}
        data-qa={props.tableqas.toolbar}
        customclasses={classnames(addCustomClasses('tableToolbar'))}
        customSearchbarClasses={classnames(addCustomClasses('tableSearchbar'))}
        paperElevation={props.paperElevation}
        extraIcons={props.extraIcons}
      />
      {
        props.showTopPagination ?
          renderTablePagination(classnames(addCustomClasses('tablePagination', css.tableTopPagination))) : <Fragment />
      }
      <MUITable
        stickyHeader={props.stickyHeader}
        className={classnames(addCustomClasses('table', css.table))}
        data-qa={props.tableqas.table}
      >
        <TableHead data-qa={props.tableqas.header} className={classnames(addCustomClasses('tableHeader'))}>
          <TableRow data-qa={props.tableqas.headerRow} className={classnames(addCustomClasses('tableHeaderRow'))}>
            {props.headers.map(header => (
              <TableHeaderCell
                key={header.label}
                header={header}
                orderby={props.orderby}
                order={props.order || 'asc'}
                tooltiplabel={props.labels.tooltiplabel}
                tooltipplacement={props.tooltipplacement}
                onRequestSort={props.onRequestSort}
                data-qa={props.tableqas.headerCell}
                customclasses={classnames(addCustomClasses('tableHeaderCell'))}
                isActionButtonCell={header.isActionButtonCell}
              />
            ))}
          </TableRow>
        </TableHead>
        <TableBody data-qa={props.tableqas.tableBody} className={classnames(addCustomClasses('tableBody'))}>
          {props.tablebodycontent}
        </TableBody>
      </MUITable>
      {
        props.showBottomPagination ?
          renderTablePagination(classnames(addCustomClasses('tablePagination'))) : <Fragment />
      }
    </Fragment>
  );
};

export default memo(Table);