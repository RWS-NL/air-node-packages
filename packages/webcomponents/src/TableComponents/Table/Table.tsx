import MUITable, { TableProps as MUITableProps } from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableHead from '@material-ui/core/TableHead';
import { TablePaginationProps as MUITablePaginationProps } from '@material-ui/core/TablePagination';
import TableRow from '@material-ui/core/TableRow';
import classnames from 'classnames';
import React, { FC, Fragment, ReactNode, memo } from 'react';
import css from './Table.scss';
import TableHeaderCell, { TableHeaderCellProps, TableHeaderProps } from '../HeaderCell/HeaderCell';
import Pagination, { PaginationProps } from '../Pagination/Pagination';
import TableToolbar, { TableToolbarProps } from '../Toolbar/Toolbar';

export type TableLabels =
  & Pick<TableToolbarProps, 'searchplaceholderlabel'>
  & Pick<PaginationProps, 'labelRowsPerPage' | 'labelPaginationOf'>
  & Pick<TableHeaderCellProps, 'tooltiplabel'>;

export interface TableQAs {
  /** Data-qa applied to the root table */
  table: string;
  /** Data-qa applied to the pagination components */
  pagination: string;
  /** Data-qa applied to the toolbar */
  toolbar: string;
  /** Data-qa applied to the header wrapper */
  header: string;
  /** Data-qa applied to the header row */
  headerRow: string;
  /** Data-qa applied to each header cell */
  headerCell: string;
  /** Data-qa applied to the table body wrapper */
  tableBody: string;
}

export interface TableCustomClasses {
  /** Custom class(es) for table */
  table?: string[];
  /** Custom class(es) for table toolbar */
  tableToolbar?: string[];
  /** Custom class(es) for table paginations */
  tablePagination?: string[];
  /** Custom class(es) for table pagination actions */
  tablePaginationActions?: string[];
  /** Custom class(es) for table header */
  tableHeader?: string[];
  /** Custom class(es) for table header row */
  tableHeaderRow?: string[];
  /** Custom class(es) for table header cell */
  tableHeaderCell?: string[];
  /** Custom class(es) for table body */
  tableBody?: string[];
  /** Custom class(es) for table toolbar searchbar */
  tableSearchbar?: string[];
}

export interface TableProps extends
  Pick<TableToolbarProps, 'onsearchclear' | 'onsearchinput' | 'paperElevation' | 'extraIcons' | 'searchdebounce'>,
  Pick<TableHeaderCellProps, 'onRequestSort' | 'orderby' | 'order' | 'tooltipplacement'>,
  Pick<MUITablePaginationProps, 'rowsPerPage' | 'rowsPerPageOptions' | 'page' | 'onChangePage' | 'onChangeRowsPerPage'>,
  MUITableProps {
  /** Headers for the table */
  headers: TableHeaderProps[];
  /** Map that defines how the headers are sorted, either by key or function when targeting nested properties */
  headermapping: Map<string, string>;
  /** The amount of rows for the table (generally the length of the data) */
  rowcount: number;
  /** Labels used in the table */
  labels: TableLabels;
  /** Data-qa's used in the table */
  tableqas: TableQAs;
  /** Content for the table */
  tablebodycontent: ReactNode;
  /** Customization CSS classes for table components */
  tablecss?: TableCustomClasses;
  /** Control whether the top bar pagination should be shown */
  showTopPagination?: boolean;
  /** Control whether the bottom bar pagination should be shown */
  showBottomPagination?: boolean;
}

/**
 * Constructs a table using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the table
 * @example
 * <Table
 *   showBottomPagination
 *   showTopPagination
 *   paperElevation={1}
 *   onsearchclear={props.handleSearchClear}
 *   onsearchinput={props.handleSearchInput}
 *   onRequestSort={props.handleRequestSort}
 *   onChangePage={props.handleChangePage}
 *   onChangeRowsPerPage={props.handleChangeRowsPerPage}
 *   tooltipplacement='bottom-start'
 *   order={props.tableOrder}
 *   orderby={props.tableOrderBy}
 *   rowsPerPage={props.tableRowsPerPage}
 *   rowsPerPageOptions={tableRowsPerPage}
 *   page={props.tablePage}
 *   headers={tableHeaders}
 *   headermapping={tableHeaderMapping}
 *   rowcount={props.data.length}
 *   labels={{
 *     labelPaginationOf: t('table.generic.pagination.of_label'),
 *     labelRowsPerPage: t('table.generic.pagination.rows_per_page'),
 *     searchplaceholderlabel: t('table.generic.searchLabel'),
 *     tooltiplabel: t('table.generic.sortTooltipLabel'),
 *   }}
 *   tableqas={{
 *     header: 'table-header',
 *     headerRow: 'table-header-row',
 *     pagination: 'table-pagination',
 *     table: 'table',
 *     toolbar: 'table-toolbar',
 *     headerCell: 'table-header-cell',
 *     tableBody: 'table-body',
 *   }}
 *   extraIcons={[
 *     {
 *       icon: <CloudDownload />,
 *       clickEvent: () => console.log('Clicked button!'),
 *       tooltipText: 'Tooltip on button',
 *     }
 *   ]}
 *   tablebodycontent={(
 *     <Fragment>
 *       {props.data
 *         .slice(props.tablePage * props.tableRowsPerPage, props.tablePage * props.tableRowsPerPage + Number(props.tableRowsPerPage))
 *         .map((item, index) => {
 *           return (
 *             <TableRow hover tabIndex={-1} key={index} data-qa='table-body-row'>
 *               <TableBodyCell content={item.name} />
 *               <TableBodyCell content={item.email} />
 *             </TableRow>
 *           );
 *         })
 *       }
 *     </Fragment>
 *   )}
 * />
 */
export const Table: FC<TableProps> = props => {
  const addCustomClasses = (component: keyof TableCustomClasses, baseClass?: string): string[] => {
    const classes: string[] = [];
    if (baseClass) classes.push(baseClass);
    if (props.tablecss && props.tablecss[component]) classes.push(...props.tablecss[component]!); // eslint-disable-line @typescript-eslint/no-non-null-assertion

    return classes;
  };

  const renderTablePagination = (customClasses: string) => {
    return (
      <Pagination
        labelRowsPerPage={props.labels.labelRowsPerPage}
        labelPaginationOf={props.labels.labelPaginationOf}
        rowsPerPageOptions={props.rowsPerPageOptions}
        rowsPerPage={props.rowsPerPage}
        page={props.page}
        count={props.rowcount}
        onChangePage={props.onChangePage}
        onChangeRowsPerPage={props.onChangeRowsPerPage}
        customClasses={customClasses}
        data-qa={props.tableqas.pagination}
      />
    );
  };

  return (
    <Fragment>
      <TableToolbar
        searchplaceholderlabel={props.labels.searchplaceholderlabel}
        onsearchinput={props.onsearchinput}
        onsearchclear={props.onsearchclear}
        searchdebounce={props.searchdebounce}
        data-qa={props.tableqas.toolbar}
        customclasses={classnames(addCustomClasses('tableToolbar'))}
        customSearchbarClasses={classnames(addCustomClasses('tableSearchbar'))}
        paperElevation={props.paperElevation}
        extraIcons={props.extraIcons}
      />
      {
        props.showTopPagination ?
          renderTablePagination(classnames(addCustomClasses('tablePagination', css.tableTopPagination))) : <Fragment />
      }
      <MUITable
        stickyHeader={props.stickyHeader}
        className={classnames(addCustomClasses('table', css.table))}
        data-qa={props.tableqas.table}
      >
        <TableHead data-qa={props.tableqas.header} className={classnames(addCustomClasses('tableHeader'))}>
          <TableRow data-qa={props.tableqas.headerRow} className={classnames(addCustomClasses('tableHeaderRow'))}>
            {props.headers.map((header, index) => (
              <TableHeaderCell
                key={index}
                header={header}
                orderby={props.orderby}
                order={props.order || 'asc'}
                tooltiplabel={props.labels.tooltiplabel}
                tooltipplacement={props.tooltipplacement}
                onRequestSort={props.onRequestSort}
                data-qa={props.tableqas.headerCell}
                customclasses={classnames(addCustomClasses('tableHeaderCell'))}
                isActionButtonCell={header.isActionButtonCell}
              />
            ))}
          </TableRow>
        </TableHead>
        <TableBody data-qa={props.tableqas.tableBody} className={classnames(addCustomClasses('tableBody'))}>
          {props.tablebodycontent}
        </TableBody>
      </MUITable>
      {
        props.showBottomPagination ?
          renderTablePagination(classnames(addCustomClasses('tablePagination'))) : <Fragment />
      }
    </Fragment>
  );
};

export default memo(Table);