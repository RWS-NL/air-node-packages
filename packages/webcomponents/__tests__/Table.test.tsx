/* eslint-disable no-console */
import { TableRow } from '@material-ui/core';
import { shallow, ShallowWrapper } from 'enzyme';
import React, { Fragment } from 'react';
import Table, { TableProps } from '../src/Table/Table';
import TableBodyCell from '../src/TableBodyCell/TableBodyCell';
import TableHeaderCell, { TableHeaderProps } from '../src/TableHeaderCell/TableHeaderCell';

interface DataForTableType {
  name: string;
  email: string;
  id: number;
}

const dataForTable: DataForTableType[] = [
  { name: 'Robin Hood', email: 'robin.hood@winked.com', id: 23456789 },
  { name: 'Darth Vader', email: 'darth.vader@thedeathstart.com', id: 9876 },
  { name: 'Kaladin Stormblessed', email: 'kaladin.stormblessed@thearmy.com', id: 567890 },
  { name: 'Steve Jobs', email: 'steve.jobs@apple.com', id: 1 }
];

const dataTableHeaderMapping: Map<string, string> = new Map(
  [
    [
      'name', 'name'
    ],
    [
      'email', 'email'
    ],
    [
      'id', 'id'
    ],
    [
      'action', 'action'
    ]
  ]
);

const tableHeaders: TableHeaderProps[] = [
  { label: Array.from(dataTableHeaderMapping.keys())[0] },
  { label: Array.from(dataTableHeaderMapping.keys())[1] },
  { label: Array.from(dataTableHeaderMapping.keys())[2], numeric: true },
  { label: Array.from(dataTableHeaderMapping.keys())[3], isActionButtonCell: true }
];

const rowsPerPage = 5;
const page = 0;
const mockOnRequestSort = jest.fn();
const mockOnSearchInput = jest.fn();
const mockOnSearchClear = jest.fn();
const mockOnChangePage = jest.fn();
const mockOnChangeRowsPerPage = jest.fn();

const propsForTable: TableProps = {
  onsearchclear: mockOnSearchClear,
  onsearchinput: mockOnSearchInput,
  onRequestSort: mockOnRequestSort,
  tooltipplacement: 'bottom-start',
  order: 'asc',
  orderby: 'name',
  rowsPerPage,
  rowsPerPageOptions: [ 2, 4, 5, 10 ],
  page,
  onChangePage: mockOnChangePage,
  onChangeRowsPerPage: mockOnChangeRowsPerPage,
  headers: tableHeaders,
  headermapping: dataTableHeaderMapping,
  rowcount: dataForTable.length,
  labels: {
    labelRowsPerPage: 'Rows per page',
    searchplaceholderlabel: 'Search...',
    tooltiplabel: 'Sort',
    labelPaginationOf: 'of',
  },
  tableqas: {
    header: 'table-header',
    headerRow: 'table-header-row',
    pagination: 'table-pagination',
    table: 'table',
    toolbar: 'table-toolbar',
    headerCell: 'table-header-cell',
    tableBody: 'table-body',
  },
  tablebodycontent: (
    <Fragment>
      {dataForTable
        .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
        .map(row => (
          <TableRow hover tabIndex={-1} key={4567893} data-qa='table-body-row'>
            <TableBodyCell content={row.name} />
            <TableBodyCell content={row.email} />
            <TableBodyCell content={row.id} />
          </TableRow>
        ))
      }
    </Fragment>
  ),
};

describe('Component Tests', () => {
  let table: ShallowWrapper;

  beforeEach(() => {
    table = shallow(<Table {...propsForTable} />);
  });

  afterEach(() => {
    mockOnRequestSort.mockRestore();
    mockOnSearchInput.mockRestore();
    mockOnSearchClear.mockRestore();
    mockOnChangePage.mockRestore();
    mockOnChangeRowsPerPage.mockRestore();
  });

  test('should request sorting on header click', () => {
    const firstColumnHeaderCell = table.find(TableHeaderCell).at(0).shallow();
    const tableSortHeader = firstColumnHeaderCell.find(`[data-qa="tableSortLabel_${tableHeaders[0].label}"]`);

    tableSortHeader.simulate('click');

    expect(mockOnRequestSort).toHaveBeenCalledWith('name');
    expect(mockOnRequestSort).toHaveBeenCalledTimes(1);
  });

  test('should not request sorting on action header click', () => {
    const lastColumnHeaderCell = table.find(TableHeaderCell).last().shallow();
    const tableSortHeader = lastColumnHeaderCell.find(`[data-qa="tableSortLabel_${tableHeaders[3].label}"]`);

    tableSortHeader.simulate('click');

    expect(mockOnRequestSort).toHaveBeenCalledTimes(0);
  });

  describe('Table Headers', () => {
    test('should render header rows', () => {
      expect(table.find('[data-qa="table-header-row"]')).toHaveLength(1);
    });

    test('should put header cells in header row with tableHeaders', () => {
      // Stub out the console errors coming from Material-UI
      const originalLogger = console.error;
      jest.spyOn(console, 'error').mockImplementation();

      const firstRow = table.find('[data-qa="table-header-row"]').first();
      const tableHeaderCell = firstRow.find(TableHeaderCell).first().shallow().find(`[data-qa="tableSortLabel_${tableHeaders[0].label}"]`);
      expect(firstRow.find(TableHeaderCell)).toHaveLength(4);
      expect(tableHeaderCell.render().text()).toBe(tableHeaders[0].label);

      // Restore console errors
      console.error = originalLogger;
    });
  });

  describe('Table Body', () => {
    test('should render body rows for first page', () => {
      expect(table.find('[data-qa="table-body-row"]')).toHaveLength(4);
    });

    test('should fill first row with all assigned data', () => {
      const firstRow = table.find('[data-qa="table-body-row"]').first();
      expect(firstRow.find(TableBodyCell)).toHaveLength(3);

      expect(firstRow.find(TableBodyCell).at(0).prop('content')).toBe(dataForTable[0].name);
      expect(firstRow.find(TableBodyCell).at(1).prop('content')).toBe(dataForTable[0].email);
      expect(firstRow.find(TableBodyCell).at(2).prop('content')).toBe(dataForTable[0].id);
    });
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const table = shallow(<Table {...propsForTable} />);
    expect(table).toMatchSnapshot();
  });

  test('Optional Props', () => {
    const table = shallow(<Table {...propsForTable} draggable showTopPagination showBottomPagination />);
    expect(table).toMatchSnapshot();
  });
});