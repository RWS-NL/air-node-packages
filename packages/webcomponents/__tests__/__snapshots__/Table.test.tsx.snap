// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Optional Props 1`] = `
<Fragment>
  <Memo(TableToolbar)
    customSearchbarClasses=""
    customclasses=""
    data-qa="table-toolbar"
    onsearchclear={[MockFunction]}
    onsearchinput={[MockFunction]}
    searchplaceholderlabel="Search..."
  />
  <Memo(TablePagination)
    count={4}
    customClasses="tableTopPagination"
    data-qa="table-pagination"
    labelPaginationOf="of"
    labelRowsPerPage="Rows per page"
    onChangePage={[MockFunction]}
    onChangeRowsPerPage={[MockFunction]}
    page={0}
    rowsPerPage={5}
    rowsPerPageOptions={
      Array [
        2,
        4,
        5,
        10,
      ]
    }
  />
  <WithStyles(ForwardRef(Table))
    className="table"
    data-qa="table"
  >
    <WithStyles(ForwardRef(TableHead))
      className=""
      data-qa="table-header"
    >
      <WithStyles(ForwardRef(TableRow))
        className=""
        data-qa="table-header-row"
      >
        <Memo(TableHeaderCell)
          customclasses=""
          data-qa="table-header-cell"
          header={
            Object {
              "label": "name",
            }
          }
          key="name"
          onRequestSort={
            [MockFunction] {
              "calls": Array [
                Array [
                  "name",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          order="asc"
          orderby="name"
          tooltiplabel="Sort"
          tooltipplacement="bottom-start"
        />
        <Memo(TableHeaderCell)
          customclasses=""
          data-qa="table-header-cell"
          header={
            Object {
              "label": "email",
            }
          }
          key="email"
          onRequestSort={
            [MockFunction] {
              "calls": Array [
                Array [
                  "name",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          order="asc"
          orderby="name"
          tooltiplabel="Sort"
          tooltipplacement="bottom-start"
        />
        <Memo(TableHeaderCell)
          customclasses=""
          data-qa="table-header-cell"
          header={
            Object {
              "label": "id",
              "numeric": true,
            }
          }
          key="id"
          onRequestSort={
            [MockFunction] {
              "calls": Array [
                Array [
                  "name",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          order="asc"
          orderby="name"
          tooltiplabel="Sort"
          tooltipplacement="bottom-start"
        />
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody))
      className=""
      data-qa="table-body"
    >
      <WithStyles(ForwardRef(TableRow))
        data-qa="table-body-row"
        hover={true}
        key="4567893"
        tabIndex={-1}
      >
        <Memo(TableBodyCell)
          content="Robin Hood"
        />
        <Memo(TableBodyCell)
          content="robin.hood@winked.com"
        />
        <Memo(TableBodyCell)
          content={23456789}
        />
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        data-qa="table-body-row"
        hover={true}
        key="4567893"
        tabIndex={-1}
      >
        <Memo(TableBodyCell)
          content="Darth Vader"
        />
        <Memo(TableBodyCell)
          content="darth.vader@thedeathstart.com"
        />
        <Memo(TableBodyCell)
          content={9876}
        />
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        data-qa="table-body-row"
        hover={true}
        key="4567893"
        tabIndex={-1}
      >
        <Memo(TableBodyCell)
          content="Kaladin Stormblessed"
        />
        <Memo(TableBodyCell)
          content="kaladin.stormblessed@thearmy.com"
        />
        <Memo(TableBodyCell)
          content={567890}
        />
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        data-qa="table-body-row"
        hover={true}
        key="4567893"
        tabIndex={-1}
      >
        <Memo(TableBodyCell)
          content="Steve Jobs"
        />
        <Memo(TableBodyCell)
          content="steve.jobs@apple.com"
        />
        <Memo(TableBodyCell)
          content={1}
        />
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableBody))>
  </WithStyles(ForwardRef(Table))>
  <Memo(TablePagination)
    count={4}
    customClasses=""
    data-qa="table-pagination"
    labelPaginationOf="of"
    labelRowsPerPage="Rows per page"
    onChangePage={[MockFunction]}
    onChangeRowsPerPage={[MockFunction]}
    page={0}
    rowsPerPage={5}
    rowsPerPageOptions={
      Array [
        2,
        4,
        5,
        10,
      ]
    }
  />
</Fragment>
`;

exports[`Snapshot Testing Required Props 1`] = `
<Fragment>
  <Memo(TableToolbar)
    customSearchbarClasses=""
    customclasses=""
    data-qa="table-toolbar"
    onsearchclear={[MockFunction]}
    onsearchinput={[MockFunction]}
    searchplaceholderlabel="Search..."
  />
  <WithStyles(ForwardRef(Table))
    className="table"
    data-qa="table"
  >
    <WithStyles(ForwardRef(TableHead))
      className=""
      data-qa="table-header"
    >
      <WithStyles(ForwardRef(TableRow))
        className=""
        data-qa="table-header-row"
      >
        <Memo(TableHeaderCell)
          customclasses=""
          data-qa="table-header-cell"
          header={
            Object {
              "label": "name",
            }
          }
          key="name"
          onRequestSort={
            [MockFunction] {
              "calls": Array [
                Array [
                  "name",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          order="asc"
          orderby="name"
          tooltiplabel="Sort"
          tooltipplacement="bottom-start"
        />
        <Memo(TableHeaderCell)
          customclasses=""
          data-qa="table-header-cell"
          header={
            Object {
              "label": "email",
            }
          }
          key="email"
          onRequestSort={
            [MockFunction] {
              "calls": Array [
                Array [
                  "name",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          order="asc"
          orderby="name"
          tooltiplabel="Sort"
          tooltipplacement="bottom-start"
        />
        <Memo(TableHeaderCell)
          customclasses=""
          data-qa="table-header-cell"
          header={
            Object {
              "label": "id",
              "numeric": true,
            }
          }
          key="id"
          onRequestSort={
            [MockFunction] {
              "calls": Array [
                Array [
                  "name",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            }
          }
          order="asc"
          orderby="name"
          tooltiplabel="Sort"
          tooltipplacement="bottom-start"
        />
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody))
      className=""
      data-qa="table-body"
    >
      <WithStyles(ForwardRef(TableRow))
        data-qa="table-body-row"
        hover={true}
        key="4567893"
        tabIndex={-1}
      >
        <Memo(TableBodyCell)
          content="Robin Hood"
        />
        <Memo(TableBodyCell)
          content="robin.hood@winked.com"
        />
        <Memo(TableBodyCell)
          content={23456789}
        />
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        data-qa="table-body-row"
        hover={true}
        key="4567893"
        tabIndex={-1}
      >
        <Memo(TableBodyCell)
          content="Darth Vader"
        />
        <Memo(TableBodyCell)
          content="darth.vader@thedeathstart.com"
        />
        <Memo(TableBodyCell)
          content={9876}
        />
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        data-qa="table-body-row"
        hover={true}
        key="4567893"
        tabIndex={-1}
      >
        <Memo(TableBodyCell)
          content="Kaladin Stormblessed"
        />
        <Memo(TableBodyCell)
          content="kaladin.stormblessed@thearmy.com"
        />
        <Memo(TableBodyCell)
          content={567890}
        />
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))
        data-qa="table-body-row"
        hover={true}
        key="4567893"
        tabIndex={-1}
      >
        <Memo(TableBodyCell)
          content="Steve Jobs"
        />
        <Memo(TableBodyCell)
          content="steve.jobs@apple.com"
        />
        <Memo(TableBodyCell)
          content={1}
        />
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableBody))>
  </WithStyles(ForwardRef(Table))>
</Fragment>
`;
